trust-quorum/gfss/src/polynomial.rs

```diff
@@ -0,0 +1,178 @@
+        // to the degree.
+        let mut inner: Vec<Gf256> =
+            rng.sample_iter(distributions::Standard).take(degree + 1).collect();
+
+        // Overwrite the constant term
+        inner[0] = constant_term;
+
+        // Ensure the highest term coefficient is nonzero.
+        //
+        // This is not a sensitive operation. It is public information that this
+        // term must not be zero, and so looping for as long as it takes to get
+        // a nonzero byte doesn't provide the attacker with any new information.
+        //
+        // However, we still want our comparison to be constant time so we use
+        // `ct_eq`.
+        while inner[degree].ct_eq(&gf256::ZERO).into() {
+            inner[degree] = rng.r#gen();
+        }
+
+        Polynomial(inner)
+    }
+
+    /// Evaluate the polynomial function at the given `x` coordinate
+    /// and return the resulting `y` coordinate.
+    ///
+    /// This is an O(n) version using Horner's method
+    pub fn eval(&self, x: Gf256) -> Gf256 {
+        let mut iter = self.0.iter().rev();
+        let mut y = *(iter.next().unwrap());
+        for coefficient in iter {
+            y = y * x + *coefficient
+        }
+        y
+    }
+}
+
+impl Display for Polynomial {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // We don't print the plus sign the first time through, and we only
+        // print it when we know there is a nonzero coefficient.
+        let mut print_plus_sign = false;
+        for (exponent, coefficient) in self.0.iter().enumerate().rev() {
+            if coefficient.ct_eq(&gf256::ZERO).into() {
+                continue;
+            }
+            if print_plus_sign {
+                write!(f, " + ")?;
+            }
+            if exponent == 0 {
+                write!(f, "{coefficient}")?;
+            } else if exponent == 1 {
+                write!(f, "{coefficient}x")?;
+            } else {
+                write!(f, "{coefficient}x^{exponent}")?;
+                print_plus_sign = true;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::gf256::Gf256;
+    use rand::thread_rng;
+    use test_strategy::{Arbitrary, proptest};
+
+    impl Polynomial {
+        /// Evaluate the polynomial function at the given `x` coordinate
+        /// and return the resulting `y` coordinate.
+        ///
+        /// This is the 0(n^2) naive method used for testing
+        fn naive_eval(&self, x: Gf256) -> Gf256 {
+            let mut y = Gf256::new(0);
+            for (exponent, coefficient) in self.0.iter().enumerate() {
+                // Safety: We only construct a `Polynomial` by passing in a
+                // `ValidThreshold` which always fits in a u8 and is the number of
+                // degrees of our polynomial + 1.
+                //
+                // We could call `try_into.unwrap()` but that adds instructions and
```

Comment:
This comment is stale, from before I implemented Horner's method and this was the main code path.